pages/slice/language-reference/slice-grammar.md

```diff
@@ -2,14 +2,587 @@
+and can be referenced with their keywords:
+
+```ebnf {% showTitle=false %}
+Primitive
+    : "bool"      | "int8"  | "uint8"  | "int16"    | "uint16"    | "int32"   | "uint32"  | "varint32"
+    | "varuint32" | "int64" | "uint64" | "varint62" | "varuint62" | "float32" | "float64" | "string"
+    | "AnyClass"
+    ;
+```
+
+For additional information on primitive types, see the [primitive type][primitive-type-guide] page.
+
+### Collection types
+
+Slice supports 2 built-in collection types: sequences and dictionaries.
+
+These types are generic over the kinds of elements they contain.
+Sequences require a single type parameter: the _element_ type.
+Dictionaries require two type parameters: the _key_ and _value_ types.
+
+Any Slice type can be specified for element and value types, but since dictionary keys must be comparable, only the following are valid key types:
+- `bool`
+- `string`
+- integral types (`int8`, `uint8`, `int16`, `uint16`, `int32`, `uint32`, `varint32`, `varuint32`, `int64`, `uint64`, `varint62`, `varuint62`)
+- enum types
+- custom types
+- compact structs whose fields are also valid key types
+
+Sequences and dictionaries can be referenced with their respective keywords (`Sequence` and `Dictionary`),
+but must have concrete types specified for their type parameters.
+
+```ebnf {% showTitle=false %}
+Sequence
+    : "Sequence" "<" TypeRef ">"
+    ;
+
+Dictionary
+    : "Dictionary" "<" TypeRef "," TypeRef ">"
+    ;
+```
+
+```slice
+Sequence<bool>
+Dictionary<uint8, string>
+```
+
+For additional information on sequences and dictionaries, see the [sequence][sequence-guide] and [dictionary][dictionary-guide] pages.
+
+### Struct types
+
+Struct definitions consist of the `struct` keyword, followed by an [identifier], and then the struct's body.
+Struct bodies consist of a list of [fields][field] wrapped in a pair of braces. These fields may be optionally separated by commas.
+Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the struct in its prelude.
+
+```ebnf {% showTitle=false %}
+Struct
+    : Prelude "compact"? "struct" Identifier "{" UndelimitedList<Field> "}"
+    ;
+```
+
+Structs also support the `compact` modifier keyword.
+A [compact struct][compact-struct-guide] cannot contain [tagged fields][tag].
+
+For additional information on structs, see the [struct][struct-guide] page.
+
+### Class types
+
+Classes can only be defined or referenced in [`Slice1`][compilation-mode-guide] mode.
+Classes can never be tagged when used as the type of a [field] or [parameter].
+
+Class definitions consist of the `class` keyword, followed by an [identifier], optionally followed by a [compact ID][compact-id-guide], optionally followed by a base class, and then the class's body.
+Compact IDs consist of a positive integer wrapped in a pair of parenthesis.
+This integer must between `0` and `2,147,483,647` (the maximum value of a signed 32bit integer).
```

Comment:
```suggestion
This integer must between `0` and `2,147,483,647` (the maximum value of a signed 32 bit integer).
```
Should this have a space?

Reply:
Looking online, the convention seems to be using a hyphen. Fixed.